tracer/src/Datadog.Trace/Debugger/Conditions/ProbeCondition.cs

```diff
@@ -0,0 +1,143 @@
+        }
+
+        public Lazy<Condition> Condition { get; set; }
+
+        public bool ShouldEvaluate { get; set; }
+
+        public string Id { get; }
+
+        // we can save this as T but it will require native changes to return MethodDebuggerState<T> instead MethodDebuggerState
+        public ScopeMember InvocationTarget { get; private set; }
+
+        public MethodPhase EvaluateAt { get; }
+
+        public string ConditionJson { get; }
+
+        public ReadOnlyCollection<ScopeMember> ScopeMembers => _scopeMembers.AsReadOnly();
+
+        public bool IsEntryMethodCondition
+        {
+            get
+            {
+                return EvaluateAt == MethodPhase.Entry;
+            }
+        }
+
+        public bool IsEndMethodCondition
+        {
+            get
+            {
+                return !IsEntryMethodCondition;
+            }
+        }
+
+        private void AddMember(string name, Type type, object value, ScopeMember.MemberType eleMemberType)
+        {
+            if (!Enum.IsDefined(typeof(ScopeMember.MemberType), eleMemberType))
+            {
+                throw new ArgumentException(nameof(eleMemberType) + "is not a valid enum value");
+            }
+
+            _scopeMembers.Add(new ScopeMember(name, type, value, eleMemberType));
+        }
+
+        public void AddArgument(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Argument);
+        }
+
+        public void AddLocal(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Local);
+        }
+
+        public void AddException(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Exception);
+        }
+
+        public void AddReturn(string name, Type type, object value)
+        {
+            AddMember(name, type, value, ScopeMember.MemberType.Return);
+        }
+
+        public void SetInvocationTarget(string name, Type type, object value)
+        {
+            InvocationTarget = new ScopeMember(name, type, value, ScopeMember.MemberType.This);
+        }
+
+        public bool Evaluate()
+        {
+            if (InvocationTarget.Type == null || string.IsNullOrEmpty(ConditionJson))
+            {
+                throw new NullReferenceException("Either invocation target is null or condition json file is null or empty");
+            }
+
+            ShouldEvaluate = false;
+            try
+            {
+                return Condition.Value.Predicate(InvocationTarget, ScopeMembers);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, $"Failed to run condition for probe {Id}. Condition was {ConditionJson}");
+                return true;
+            }
+        }
+
+        public void SetShouldEvaluate(MethodPhase methodPhase)
```

Comment:
We're already exposing `EvaluateAt`, so wouldn't it be simpler and preferable to just have the consume deduce whether we should evaluate or not using that, rather than exposing this somewhat chatty API? In other words, can we remove `SetShouldEvaulate` and `ShouldEvaluate`, and just use `EvaluateAt` directly?

Reply:
Maybe we can think on a way but for now we can't because EvaluateAt can tell us about entry or end but from both phases we have to capture arguments and locals. Do you a suggestion how to do that?